java-frontend/src/main/java/org/sonar/java/caching/ContentHashCache.java

```diff
@@ -0,0 +1,122 @@
+ */
+package org.sonar.java.caching;
+
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.cache.ReadCache;
+import org.sonar.api.batch.sensor.cache.WriteCache;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class ContentHashCache {
+
+  private static final Logger LOG = Loggers.get(ContentHashCache.class);
+  private static final String CONTENT_HASH_KEY = String.format("java:contentHash:%s:", FileHashingUtils.HASH_ALGORITHM);
+  private static final String HASH_COMPUTE_FAIL_MSG = "Failed to compute content hash for file %s";
+
+  private ReadCache readCache;
+  private WriteCache writeCache;
+  private final boolean enabled;
+
+  public ContentHashCache(SensorContext context) {
+    CacheContextImpl cacheContext = CacheContextImpl.of(context);
+    enabled = cacheContext.isCacheEnabled();
+
+    if (enabled) {
+      readCache = context.previousCache();
+      writeCache = context.nextCache();
+    }
+  }
+
+  public boolean hasSameHashCached(InputFile inputFile) {
+    if (!enabled) {
+      if (!inputFile.status().equals(InputFile.Status.SAME)) {
+        LOG.trace(() -> "Cache is disabled. File status is: " + inputFile.status() + ". File can be skipped.");
+        writeToCache(inputFile);
+        return false;
+      }
+      LOG.trace(() -> "Cannot read cached hashes when the cache is disabled (" + inputFile.key() + ").");
+      return false;
+    }
+    String cacheKey = getCacheKey(inputFile);
+    try {
+      LOG.trace("Reading cache for the file {}", inputFile.key());
+      byte[] cachedHash = readCache.read(cacheKey).readAllBytes();
+      byte[] fileHash = FileHashingUtils.inputFileContentHash(inputFile);
+      boolean isHashEqual = MessageDigest.isEqual(fileHash, cachedHash);
+      if (isHashEqual) {
+        copyFromPrevious(inputFile);
+      } else {
+        writeToCache(inputFile);
+      }
+      return isHashEqual;
+    } catch (IllegalArgumentException e) {
+      LOG.trace(String.format("Could not find key %s in the cache", cacheKey));
+      writeToCache(inputFile);
+    } catch (IOException | NoSuchAlgorithmException e) {
+      LOG.trace(String.format(HASH_COMPUTE_FAIL_MSG, inputFile.key()));
+    }
+    return false;
+  }
+
+  public boolean contains(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace(() -> "Cannot lookup cached hashes when the cache is disabled (" + inputFile.key() + ").");
+      return false;
+    }
+    return readCache.contains(getCacheKey(inputFile));
+  }
+
+  public boolean writeToCache(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace(() -> "Cannot write hashes to the cache when the cache is disabled (" + inputFile.key() + ").");
+      return false;
+    }
+    if (writeCache != null) {
```

Comment:
In practice, I think it is enough to check that the cache is enabled. Even if the cache is disabled, the documentation tells us that we should [not receive null from the sensorContext](https://github.com/SonarSource/sonar-plugin-api/blob/38fbf9cadc373fd5853e20e23dcc8cd6c3338123/plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java#L241).